module/skill/QuinticWalk/src/QuinticWalk.cpp

```diff
@@ -0,0 +1,383 @@
+
+        // Change goals from support foot based coordinate system to trunk based coordinate system
+        // Trunk {t} from support foot {s}
+        Eigen::Isometry3f Hst;
+        Hst.linear()      = EulerIntrinsicToMatrix(thetaST);
+        Hst.translation() = rTSs;
+
+        // Flying foot {f} from support foot {s}
+        Eigen::Isometry3f Hsf;
+        Hsf.linear()      = EulerIntrinsicToMatrix(thetaSF);
+        Hsf.translation() = rFSs;
+
+        // Support foot {s} from trunk {t}
+        const Eigen::Isometry3f Hts = Hst.inverse();
+
+        // Flying foot {f} from trunk {t}
+        const Eigen::Isometry3f Htf = Hts * Hsf;
+
+        // Get desired transform for left foot {l}
+        const Eigen::Isometry3f Htl = walk_engine.get_footstep().is_left_support() ? Hts : Htf;
+
+        // Get desired transform for right foot {r}
+        const Eigen::Isometry3f Htr = walk_engine.get_footstep().is_left_support() ? Htf : Hts;
+
+        // ****DIRECTOR MOTION****
+        const NUClear::clock::time_point time = NUClear::clock::now() + Per<std::chrono::seconds>(UPDATE_FREQUENCY);
+
+        //  Legs
+        auto left_leg   = std::make_unique<LeftLegIK>();
+        auto right_leg  = std::make_unique<RightLegIK>();
+        left_leg->time  = time;
+        right_leg->time = time;
+        left_leg->Htl   = Htl.cast<double>().matrix();
+        right_leg->Htr  = Htr.cast<double>().matrix();
+        // Arms
+        auto left_arm  = std::make_unique<LeftArm>();
+        auto right_arm = std::make_unique<RightArm>();
+
+        // Loop to set the servo states
+        for (int id = 0; id < ServoID::NUMBER_OF_SERVOS; ++id) {
+            // Set the legs
+            if ((id >= ServoID::R_HIP_YAW) && (id < ServoID::HEAD_YAW)) {
+                if (id % 2 == 0) {
+                    // right legs
+                    right_leg->servos[id] = ServoState(current_config.jointGains[ServoID(id)], 100);
+                }
+                else {
+                    // left legs
+                    left_leg->servos[id] = ServoState(current_config.jointGains[ServoID(id)], 100);
+                }
+            }
+            // Set the arms
+            if (id < ServoID::R_HIP_YAW) {
+                if (id % 2 == 0) {
+                    // right arms
+                    right_arm->servos[id] = ServoCommand(time,
+                                                         current_config.arm_positions[ServoID(id)].second,
+                                                         ServoState(current_config.jointGains[ServoID(id)], 100));
+                }
+                else {
+                    // left arms
+                    left_arm->servos[id] = ServoCommand(time,
+                                                        current_config.arm_positions[ServoID(id)].second,
+                                                        ServoState(current_config.jointGains[ServoID(id)], 100));
+                }
+            }
+        }
```

Comment:
This has a very high complexity.

Some ideas
- Change the config joint gains into ServoStates (since you know the torque is 100 anyway) then plug in the config here rather than creating ServoStates
- move the comments at the end of the `if (...) {` lines eg `if (id % 2 == 0) { // right arms`
- Remove the second lot of if else depth by combining the if statements into `if (id < ServoID::R_HIP_YAW && id % 2 == 0)`
- Change the for to `id < ServoID::NUMBER_OF_SERVOS - 2` and don't both checking for the head in the first if. Then it'll probably be simple enough to remove the first lot of nested ifs as well.

Reply:
```suggestion
        for (int id = 0; id < ServoID::NUMBER_OF_SERVOS - 2; id++) {
            // Set the legs
            if ((id >= ServoID::R_HIP_YAW) && (id % 2 == 0)) { // right legs
                right_leg->servos[id] = current_cfg.servo_state[ServoID(id)];
            }
            else if ((id >= ServoID::R_HIP_YAW) && (id % 2 == 1)) { // left legs
                left_leg->servos[id] = current_cfg.servo_state[ServoID(id)];
            }
            else if ((id < ServoID::R_HIP_YAW) && (id % 2 == 0)) { // right arms
                right_arm->servos[id] = ServoCommand(time,
                                                current_config.arm_positions[ServoID(id)].second,
                                                current_cfg.servo_state[ServoID(id)]);
            }
            else if ((id < ServoID::R_HIP_YAW) && (id % 2 == 1)) { // left arms
                left_arm->servos[id] = ServoCommand(time,
                                                current_config.arm_positions[ServoID(id)].second,
                                                current_cfg.servo_state[ServoID(id)]);
            }
        }
```